### ISSUE TITLE
derive(SystemParam) fails when using associated types

### ISSUE THREAD

**Issue by SamPruden** [bevyengine/bevy#4801]
## Bevy version
0.7 and master

## What you did

```rust
use bevy::ecs::system::{SystemParam, StaticSystemParam};

#[derive(SystemParam)]
pub struct MyParam<'w, 's, P: MyParamTrait<'w, 's>> {
  inner: StaticSystemParam<'w, 's, P::Inner>,
}

pub trait MyParamTrait<'w, 's> {
  type Inner: SystemParam;
}
```

## What you expected to happen

I think this should compile fine according to the docs, even accounting for the doc update in #4787 yesterday, although it's not a clear cut case. I would certainly like it to work.

## What actually happened
```
error[E0261]: use of undeclared lifetime name `'w`
 --> src\constraints\constraints_param.rs:4:44
  |
3 | #[derive(SystemParam)]
  |          - help: consider introducing lifetime `'w` here: `'w,`
4 | pub struct MyParam<'w, 's, P: MyParamTrait<'w, 's>> {
  |                                            ^^ undeclared lifetime

error[E0261]: use of undeclared lifetime name `'s`
 --> src\constraints\constraints_param.rs:4:48
  |
3 | #[derive(SystemParam)]
  |          - help: consider introducing lifetime `'s` here: `'s,`
4 | pub struct MyParam<'w, 's, P: MyParamTrait<'w, 's>> {
  |                                                ^^ undeclared lifetime
```

## Additional information

I've had a quick go at `cargo expand`ing and manually fixing, but it's proving to be a bit beyond my current levels of Rust lifetime magic experience so far, assuming that a fix is even possible.

It seems that `'w` and `'s` don't get properly routed into the `SystemParamState` impl, but the simple fix for that then runs into problems with things not being `'static`. Adding `'static` bounds everywhere runs into

```
`MyParamTrait` is not general enough
`P` must implement `MyParamTrait<'0, '_>`, for some specific lifetime `'0`...
...but it actually implements `MyParamTrait<'w, 's>
```

I'm not sure whether this should be considered a bug, or whether I'm just trying to do something that's out of scope.

**Comment by alice-i-cecile** [comment]
Is this still failing on `main`?

**Comment by alice-i-cecile** [comment]
I don't see any reason why it shouldn't work. It's just a matter of wrangling the macro to handle associated types properly.

**Comment by DJMcNab** [comment]
So the error here is terrible, but the situation you're running into with `'static` is expected

Your trait's associated item needs to be `'static`, because we need to be able to reference it in a `'static` context.

One way to write the bound you might want is `P: for<'world, 'state> MyParamState<'world, 'state>`. Note that in this case, `'world` and `'state` are different to `'w` and `'s`.

**Comment by SamPruden** [comment]
Thanks @DJMcNab, although I think that's not quite right as `MyParamState` is a struct and cannot appear on the RHS of a bound like that. I assume that you meant `MyParamTrait`?

That may be the right technique for a manual implementation, but that doesn't seem to work for the `#[derive(SystemParam)]` case if that's what you were suggesting. That still produces an impressive spew of "may not live long enough"s, on both 0.7 and `main`.

I don't need to solve this urgently as I have a functional simpler solution for now and what I was trying to do was probably an abuse of Bevy's features anyway (dynamically created generic `Resource`s to act as backing storage for a custom `SystemParam` and associated `Access`, to save on intermediary data structures), although I may come back to this requirement when multithreading things.

**Comment by DJMcNab** [comment]
Hmm, I get a different error for:
```rust
use bevy::ecs::system::{SystemParam, StaticSystemParam};

#[derive(SystemParam)]
pub struct MyParam<'w, 's, P: for<'world, 'state> MyParamTrait<'world, 'state>> {
  inner: StaticSystemParam<'w, 's, P::Inner>,
}

pub trait MyParamTrait<'w, 's> {
  type Inner: SystemParam + 'static;
}
```
which makes sense

```rust
use bevy::ecs::system::{StaticSystemParam, SystemParam, SystemParamItem};

#[derive(SystemParam)]
pub struct MyParam<'w, 's, P: MyParamTrait + 'static> {
    _inner: StaticSystemParam<'w, 's, P::Inner>,
}

pub trait MyParamTrait {
    type Inner: SystemParam + 'static;
    fn do_thing_with(inner: SystemParamItem<Self::Inner>);
}
```
Compiles fine. Your issue is that your `MyParamTrait` is generic over lifetimes, when what you want is some way to refer to the inner parameter seperately to the lifetimes it's called with. Having the lifetimes be on the function (inferred in this case) gives you that.

**Comment by SamPruden** [comment]
Ah, no I think I actually want `MyParamTrait` to be generic over the lifetimes because I use them in `Inner`.

I'm messing around with what amounts to a weird custom API similar to `Query`, but over something other than `Component`s. I think there's a good chance that (ab)using `Resource`s in this way is the wrong approach to solve my problem (I have other parallel prototypes which are probably better), but I'm still curious whether I can technically get this to work.

Impls will look something like
```rust
impl<'w, 's, C1: Component, C2: Component> MyParamTrait<'w, 's> for (&C1, &C2) {
    type Inner = (ResMut<'w, MyRes<C1>>, ResMut<'w, MyRes<C2>>);
}
```

So I think I do need to pipe at least `'w` into here to pass onto `ResMut`, don't I? I included `'s` to make the issue more general.

Apologies if this is simple incompetence on my part! I'm still in the learning stages of Rust, and some of the more complex parts of the lifetime trait interactions aren't completely intuitive to me yet.

**Comment by alice-i-cecile** [comment]
I'm going to close this as "not our bug", but feel free to keep discussing :)

**Comment by SamPruden** [comment]
Okay, so the conclusion is that this isn't meant to be possible then? I know that what I'm trying to do is weird, and probably inadvisable, but I think it would be nice if this pattern could be handled. I'm happy to accept that it's out of scope, though.

**Comment by alice-i-cecile** [comment]
I don't think it's out of scope, I just think that there's more Rust twiddling needed to get your particular pattern to play nicely with our traits.

I'm open to being convinced, but I'm not sure there's anything we could change in this derive macro that would fix the example initially given.

**Comment by DJMcNab** [comment]
The pattern we use through `bevy`'s own code is the one demonstrated above: you make sure you have a `'static` systemparam, and then use `SystemParamItem` in your trait to access it (thus re-introducing the required lifetimes). Like I say, your trait 'cannot' be generic (unless it is also generic over the `SystemParam` I guess), because you need a `'static` version of the parameter to be able to use it generically.

**Comment by SamPruden** [comment]
So I just looked back at this section of code and it turns out I've been an absolute doofus. I was completely overcomplicating things. I somehow got it into my head that a `SystemParam` containing resource fields needed to do `Res<'w, R>`, and completely neglected to even try `Res<'static, R>`, which actually seems to work just fine. I made this so much harder than it needed to be, and what I was trying to do is actually trivial.

Thanks for both of your help, and sorry for wasting your time. I promise that I'm occasionally more competent than this!
